### PR TITLE
ARROW-683: [C++/Python] Refactor to make Date32 and Date64 types for new metadata. Test IPC roundtrip

### PR DESCRIPTION
**Pull request by wesm**
Maintains existing Python behavior (datetime.date getting converted to milliseconds)

### CODE REVIEW

**Comment by wesm** [comment]
@kou if you could look at the changes under c_glib and make sure this makes sense (per the ax_cxx_compile_stdcxx_11 issue)

**Comment by kou** [comment]
You don't need to add `c_glib/macros/*.m4`. They are generated by `autogen.sh`.
`configure.ac` change isn't also needed.

I think that your autoconf-archive is old.

You can just add the following change for `c_glib/arrow-glib/type.cpp`:

```diff
-  case arrow::Type::type::DATE:
+  case arrow::Type::type::DATE32:
```

It will pass tests on Travis CI.

I'll add `DATE64` support later.

**Comment by wesm** [comment]
OK, I'll remove them, I just wanted to check that I could compile arrow-glib locally

**Comment by wesm** [comment]
How do I update my autoconf-archive? From searching on the internet, this particular issue seems to occur quite frequently

**Comment by wesm** [comment]
`apt-get install autoconf-archive` seems to have solved the problem. I'm adding this to the README

**Comment by wesm** [comment]
@xhochy this is ready to be merged -- I created PARQUET-919 for fixing parquet_arrow separately

**Comment by kou** [comment]
> apt-get install autoconf-archive seems to have solved the problem. I'm adding this to the README

Thanks!

File: cpp/src/arrow/ipc/json-internal.cc [hunk 1]
```diff
@@ -303,7 +300,10 @@ class JsonSchemaWriter : public TypeVisitor {
 
   Status Visit(const BinaryType& type) override { return WriteVarBytes("binary", type); }
 
-  Status Visit(const DateType& type) override { return WritePrimitive("date", type); }
+  // TODO
```
Comment (xhochy) [inline cpp/src/arrow/ipc/json-internal.cc]: Is there still something to do here?

**Review by xhochy — COMMENTED** [review]
LGTM but there are two `// TODO` statements which may need addressing

File: cpp/src/arrow/ipc/json-internal.cc (hunk 1 above)
Comment (wesm) [inline cpp/src/arrow/ipc/json-internal.cc]: Testing / metadata read/write for the JSON format -- ARROW-620

**Pull request closed without merging**